### ISSUE TITLE
Speed up processing

### ISSUE THREAD

**Issue by systems-rebooter** [philipmat/discogs-xml2db#65]
Hi guys, 

What's the best ideas to speed up processing with discogs-xml2db? 

Is it possilbe to run it in parallel mode? 
Right now it occupies only 1 CPU core and processing latest 20G xml  ~3 hours (for mongo file-dump)

```
$ time discogsparser.py -i -o mongo -p "file:///HDD1/2015-06" /HDD2/2015-06/discogs_20150601_releases.xml

real    191m27.478s
user    189m22.749s
sys     2m0.714s
```

![screen shot 2015-06-19 at 12 57 46](https://cloud.githubusercontent.com/assets/8888190/8251744/c35d4844-1687-11e5-98b7-7d5790cbf8c2.png)

I am playing with http://www.gnu.org/software/parallel/ now, but maybe there are some other options.

Also i've tried to omit building md5 hashes, to speed things up a bit,  but overall processing time didn't changed much. 

Any other suggestions are more then welcomed!! 

Thanks!

**Comment by ijabz** [comment]
Hopefully this approach will pay dividends https://github.com/philipmat/discogs-xml2db/issues/35

**Comment by systems-rebooter** [comment]
Wow, @ijabz that was FAST! While I edited my comment, you already answered it :) Thanks!

**Comment by ijabz** [comment]
Thanks, but my comment is now incorrect as it only refers to the postgres dump not the mongo dump. I don't if a similar approach is possible for Mongo ?

**Comment by systems-rebooter** [comment]
Why not? Batch processing is known feature for RDBMs as well as NoSQL databases..

**Comment by ijabz** [comment]
I know its valid for Postgres but I meant is there a postgres copy like command for Mongo ?
We dont store as json, but as conventional tables in Postgres, once you start storing json in relational dbs I think you lose alot of the power and flexibility of pure sql.

**Comment by systems-rebooter** [comment]
Importing to database is fairly quick, for me, so no issue here. Bottleneck is rather on `xml --> json` conversion side.I've made some tests earlier with time measurements. They look like the following. 

![screen shot 2015-06-19 at 15 59 43](https://cloud.githubusercontent.com/assets/8888190/8253900/ad956310-169c-11e5-98b4-3fffdd36b212.png)

With upgraded hardware I'm expecting import time to drop 20 --> 10 min for 20G release dataset on mongo side, but `xml --> json` is still taking around 3 hours to complete.

**Comment by ijabz** [comment]
When you say 'Importing to database is fairly quick, for me' are you talking about Mongo or Postgres, Im talking about Postgres

**Comment by redapple** [comment]
Hello everyone,

I wanted to share some work I've been doing lately on parsing XML dumps from discogs, writing CSV files and ingesting those with PostgreSQL's `COPY FROM`.
I'm not sure how much of it can be merged into the current codebase, as [I changed the schema](https://github.com/redapple/discogs-xml2db/blob/speedup/speedup/sql/CreateTables.sql) a bit to normalize as much as I could (since my ultimate goal is to build a graph DB clone of discogs data).

All the work is in the ["speedup" branch of my fork](https://github.com/redapple/discogs-xml2db/tree/speedup/speedup):
- the parser uses `lxml` instead of `xml.sax` (see [speedup/parser.py](https://github.com/redapple/discogs-xml2db/blob/speedup/speedup/parser.py))
- the CSV exporter writes to various .csv files, 1 file per table in the schema (see [speedup/exporter.py](https://github.com/redapple/discogs-xml2db/blob/speedup/speedup/exporter.py)) ; it can also bz2-compress on demand
- there's an importer heavily inspired by @lalinsky 's [mbslave](https://bitbucket.org/lalinsky/mbslave/)

Here are some timing logs of the different operations on the `20170401` dumps.

creating bz2 compressed files from the `.xml.gz` dumps took roughly 2 hours (the progress bar does not show 100% for all entities, but everything was processed, it's just that I use count estimates for this):
```
$ time python speedup/exporter.py --bz2 --export release --export master --export label --export artist /home/paul/dl/discogs/20170401/ `pwd`/testbz2
Processing    releases:  99%|████████████████████████████████████████▊| 8336521/8382828 [1:43:00<00:34,  1348.82 releases/s]
Processing     masters:  95%|███████████████████████████████████████▋ | 1134845/1200000 [  03:51<00:13,  4909.73 masters/s]
Processing      labels: 100%|█████████████████████████████████████████| 1028767/1032884 [  01:29<00:00, 11491.94 labels/s]
Processing     artists: 100%|████████████████████████████████████████▊| 4956240/4973243 [  07:36<00:01, 10852.05 artists/s]

real	116m0.327s
user	115m44.180s
sys	0m10.444s

$ ls -lAh testbz2/
total 2,7G
-rw-rw-r-- 1 paul paul 6,1M avril 14 14:51 artist_alias.csv.bz2
-rw-rw-r-- 1 paul paul  84M avril 14 14:51 artist.csv.bz2
-rw-rw-r-- 1 paul paul  17M avril 14 14:51 artist_namevariation.csv.bz2
-rw-rw-r-- 1 paul paul 8,3M avril 14 14:51 artist_url.csv.bz2
-rw-rw-r-- 1 paul paul 9,9M avril 14 14:51 group_member.csv.bz2
-rw-rw-r-- 1 paul paul  20M avril 14 14:43 label.csv.bz2
-rw-rw-r-- 1 paul paul 1,7M avril 14 14:43 label_url.csv.bz2
-rw-rw-r-- 1 paul paul 8,9M avril 14 14:42 master_artist.csv.bz2
-rw-rw-r-- 1 paul paul  18M avril 14 14:42 master.csv.bz2
-rw-rw-r-- 1 paul paul 4,7M avril 14 14:42 master_genre.csv.bz2
-rw-rw-r-- 1 paul paul 6,4M avril 14 14:42 master_style.csv.bz2
-rw-rw-r-- 1 paul paul  53M avril 14 14:42 master_video.csv.bz2
-rw-rw-r-- 1 paul paul 275M avril 14 14:38 release_artist.csv.bz2
-rw-rw-r-- 1 paul paul 100M avril 14 14:38 release_company.csv.bz2
-rw-rw-r-- 1 paul paul 359M avril 14 14:38 release.csv.bz2
-rw-rw-r-- 1 paul paul  33M avril 14 14:38 release_format.csv.bz2
-rw-rw-r-- 1 paul paul  29M avril 14 14:38 release_genre.csv.bz2
-rw-rw-r-- 1 paul paul  93M avril 14 14:38 release_identifier.csv.bz2
-rw-rw-r-- 1 paul paul  90M avril 14 14:38 release_label.csv.bz2
-rw-rw-r-- 1 paul paul  40M avril 14 14:38 release_style.csv.bz2
-rw-rw-r-- 1 paul paul 310M avril 14 14:38 release_track_artist.csv.bz2
-rw-rw-r-- 1 paul paul 766M avril 14 14:38 release_track.csv.bz2
-rw-rw-r-- 1 paul paul 385M avril 14 14:38 release_video.csv.bz2
```

And importing these files into a clean database, WITHOUT any indices nor FKs, took a bit under 30 minutes:
```
$ ./discogs-psql.py <sql/CreateTables.sql
CREATE TABLE
...
CREATE TABLE

$ time ./discogs-import.py `ls ../testbz2/*.csv.bz2`
Importing data from /home/paul/src/discogs-xml2db/testbz2/artist_alias.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/artist.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/artist_namevariation.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/artist_url.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/group_member.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/label.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/label_url.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/master_artist.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/master.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/master_genre.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/master_style.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/master_video.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/release_artist.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/release_company.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/release.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/release_format.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/release_genre.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/release_identifier.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/release_label.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/release_style.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/release_track_artist.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/release_track.csv.bz2
Importing data from /home/paul/src/discogs-xml2db/testbz2/release_video.csv.bz2

real	25m47.296s
user	12m38.876s
sys	0m23.836s
```

Here are the table sizes in the end:

```
$ ./discogs-psql.py
psql (9.5.6)
SSL connection (protocol: TLSv1.2, cipher: ECDHE-RSA-AES256-GCM-SHA384, bits: 256, compression: off)
Type "help" for help.

discogs=> \dt+
                            List of relations
 Schema |         Name         | Type  |  Owner  |  Size   | Description
--------+----------------------+-------+---------+---------+-------------
 public | artist               | table | discogs | 433 MB  |
 public | artist_alias         | table | discogs | 43 MB   |
 public | artist_namevariation | table | discogs | 118 MB  |
 public | artist_url           | table | discogs | 56 MB   |
 public | group_member         | table | discogs | 49 MB   |
 public | label                | table | discogs | 104 MB  |
 public | label_url            | table | discogs | 12 MB   |
 public | master               | table | discogs | 80 MB   |
 public | master_artist        | table | discogs | 54 MB   |
 public | master_genre         | table | discogs | 62 MB   |
 public | master_style         | table | discogs | 74 MB   |
 public | master_video         | table | discogs | 247 MB  |
 public | release              | table | discogs | 1385 MB |
 public | release_artist       | table | discogs | 2138 MB |
 public | release_company      | table | discogs | 1051 MB |
 public | release_format       | table | discogs | 500 MB  |
 public | release_genre        | table | discogs | 460 MB  |
 public | release_identifier   | table | discogs | 679 MB  |
 public | release_label        | table | discogs | 534 MB  |
 public | release_style        | table | discogs | 530 MB  |
 public | release_track        | table | discogs | 4539 MB |
 public | release_track_artist | table | discogs | 2695 MB |
 public | release_video        | table | discogs | 1923 MB |
(23 rows)

discogs=> SELECT schemaname,relname,n_live_tup
  FROM pg_stat_user_tables
  ORDER BY relname;
 schemaname |       relname        | n_live_tup
------------+----------------------+------------
 public     | artist               |    4957096
 public     | artist_alias         |     923081
 public     | artist_namevariation |    2277212
 public     | artist_url           |     737404
 public     | group_member         |     949763
 public     | label                |    1028767
 public     | label_url            |     162487
 public     | master               |    1134845
 public     | master_artist        |    1332107
 public     | master_genre         |    1463142
 public     | master_style         |    1715908
 public     | master_video         |    1587863
 public     | release              |    8333482
 public     | release_artist       |   38704450
 public     | release_company      |    9635160
 public     | release_format       |    8503079
 public     | release_genre        |   10857169
 public     | release_identifier   |   10113252
 public     | release_label        |    9457178
 public     | release_style        |   12279053
 public     | release_track        |   73825301
 public     | release_track_artist |   50125710
 public     | release_video        |   12190477
(23 rows)

discogs=> SELECT COUNT(*) FROM release;
  count
---------
 8336521
(1 row)

discogs=> SELECT COUNT(*) FROM artist;
  count
---------
 4956240
(1 row)
```
 (ftr, I'm using a Intel(R) Core(TM) i5-2450M, no SDD, with Python 3.5.2 on Ubuntu)

**Comment by philipmat** [comment]
This is incredible work, Paul. Thank you for it!

Would you mind submitting a PR so we can look it over and test it a bit on our respective boxes?

________________________________

**Comment by redapple** [comment]
Thanks for the feedback @philipmat .
What kind of PR do you have in mind? right now, I have not touched any of the current codebase, and everything is in a `speedup/` folder.
So if you'd want PRs against current files and current db schema, I need to work on it more.

**Comment by philipmat** [comment]
Well, with the awesome results you've been getting, we sure want it :)

________________________________

**Comment by vasyl-shumskyi** [comment]
@redapple nice and sweet!

**Comment by ijabz** [comment]
Great redapple I have just seen this, it is the approach that I always wanted to do but I dont know python really. So has anyone progressed  with this in the last three years or not ?

**Comment by ijabz** [comment]
I have spent the last working with Redapples speedup branch.
In summary it continues to work with the latest Discogs import and works very well. It does require Python 3 instead of Python 2 but was easy enough to get working even for someone (me) with very limited Python knowledge, and works very quickly- it is the correct approach.
However there are quite a few changes made to the database that is created, these are either cosmetic changes (renaming of tables/column so more consistent) or more substantial changes such as new fields and  using new tables rather than array fields. These changes to the database are definently an improvement but I expect any application currently using discogsxml2db would need to make some changes if Redapples speedup branch was merged in.

There are currently three regressions - artist_images, release_images and release status data are not now being captured. 

So I am now committed to using Redapples branch, should it be merged into discogs-xml2db ? The import is so much quicker, the database is not backwards compatible but it is better designed so I tend to think so but that is your call philipmat. If you think it should be merged in I do think the missing data should be fixed first, and it would also make sense if we moved the whole project to require Python3 rather Python 2 to make things simpler for everyone.

Summary of Database Changes that I encountered
---------------------------------------------------
RELEASES_LABELS table now RELEASE_LABEL
RELEASES_FORMATS table now RELEASE_FORMAT
RELEASES_ARTISTS table now RELEASE_ARTIST
TRACKS_ARTISTS table now RELEASE_TRACK_ARTIST
TRACK table now RELEASE_TRACK table
RELEASE_ARTISTS,JOIN_RELATION now RELEASE_ARTIST.JOIN_STRING
RELEASE_TRACK_ARTIST,JOIN_RELATION now RELEASE_TRACK_ARTIST.JOIN_STRING
RELEASE_TRACK_ARTIST now use TMP_TRACK_ID to match to tmp_track in RELEASE_TRACK
RELEASE_FORMAT.FORMAT_NAME now RELEASE_FORMAT.NAME
RELEASE_FORMAT.POSITION no longer exists but can use ID field to preserve order when release has multiple formats.
LABEL.CONTACTINFO now LABEL.CONTACT_INFO
LABEL.PARENT_LABEL now LABEL.PARENT_NAME
LABEL has new PARENT_ID field
RELEASE_LABEL has extra fields
Aliases now in ARTIST_ALIAS table
TRACKS_EXTRA_ARTISTS now in TRACK_ARTIST table with EXTRA flag
RELEASES_EXTRA_ARTISTS now in RELEASE_TRACK_ARTIST table with EXTRA flag
RELEASE.GENRES now in own RELEASE_GENRE table
RELEASE.STYLES now in own RELEASE_STYLE table
RELEASE.BARCODE now in RELEASE_IDENTIFIER table 
ARTIST.ANV fields now in ARTIST_VARIATION table
ARTIST.URL fields now in ARTIST_URL table

**Comment by philipmat** [comment]
I vote yes. Python 3 is not an impediment, quite the contrary - it’s desirable to me.

**Comment by ijabz** [comment]
Can you make any changes to existing code so it works with Python3
I can then essentially copy speedup folder to main folder and issue a pull request (unless RedApple wants to)

**Comment by philipmat** [comment]
Can we not just issue the PR against the current code and effectively overwrite what's already in the repo with the new code?

**Comment by ijabz** [comment]
Yes we can do that, and I will create a PR for that soon. However it only replaces the Postgres Db Import, so if the Mongo and Json exporters still reply on Python2 and this relies on Python 3 is that going to be a problem ?

**Comment by ijabz** [comment]
So I have locally fixed the three regressions ( artist_images, release_images and release status data missing). So Im going to start work on a PR to replace the existing postgres code.

**Comment by philipmat** [comment]
If you push the branch, I have a couple of meetings in the next few days I don't have to pay too much attention to 😄 and I could take a swing at making the mongo and json exporters at least py3 compatible.

**Comment by ijabz** [comment]
ok Im working on it now, but will probably be tomorrow before I have something that actually works

**Comment by ijabz** [comment]
okay I have pushed a new commit - https://github.com/philipmat/discogs-xml2db/commit/9e26e95ea72913e39d586f585d0d5d473e2d7c36

Code is in own selfcontained speedup folder, works for me. Added an README.md file, still uses get_latest_dumps.sh in parent folder to download data dumps but does expect them to be unzipped.

**Comment by ijabz** [comment]
philipmat, have you had a chance to try this yet ?

**Comment by berz** [comment]
I tried the code in the speedup directory and it truly lives up to its name being a couple times faster than the original code. Thanks to everyone involved.
Additionally it did require significantly less disk space since both the dumps and csv files are  compressed.
Lastly it allows me to import the data straight into mysql. The way I do this now is to use discogs-xml2db to import the data into postgres, dump the database and import it into mysql. The speedup script saves me a step as the csv's files it produces can be imported into mysql. I did this in about 10 lines of bash so adding support for mysql should be easy.

I hope this can become the default approach soon.

**Comment by ijabz** [comment]
Hi berz, great that someone else has tried it.
The main difficulty with moving out of speedup and into top level is doing in such a way as to not break the other importers for mongo/.couch dbs etc (which I have no experience ore interest in)

**Comment by berz** [comment]
Adding support for mongodb and couchdb should not be difficult. The speedup code generates csv files, which is a widely used format most databases can import.
- MongoDB seems to support this out of the box with their mongoimport tool. The docs can be found [here](https://docs.mongodb.com/manual/reference/program/mongoimport/).
- CouchDB only supports json out of the box, but there are programs that make it easy to import csv. [this tutorial](https://medium.com/codait/simple-csv-import-for-couchdb-71616200b095) explains how to do this using [couchimport](https://github.com/glynnbird/couchimport).

It seems that using these databases should still work after merging the speedup code. 
The people using these will have to change their workflow or application. But postgres users will have to change their application too because the database scheme changed (for the better). 
Even though it breaks compatibility with previous versions I think merging this code is worth it as the new code is simpler, faster and makes it easy to add support for other databases. 

I do not use either mongodb or couchdb so I am not going to put in the effort to test or automate importing to these databases. I would leave that to the people using them. 
However if this gets merged I will be using it to import into mysql. I could send a pull request adding mysql support, which seems to be a feature other people have asked for.

**Comment by ijabz** [comment]
The speedup code **is already merged** into the master branch, its just just that it is a speedup subfolder rather than the main location, like you i dont use mongodb or couchdb. Ive got the speedup code working but really its up to @philipmat to reorganize the code so its in top folder if he wants to.
A pull request for mysql, would be useful, thanks.

**Comment by berz** [comment]
I added support for importing into mysql to the speedup code. My code can be found [here](https://github.com/berz/discogs-xml2db/tree/mysql). I opened a pull request for this (#98), could you have a look at it?

**Comment by ijabz** [comment]
Im merged your work, I am going to close this now, we can create issues for specific issues as needed
